### PR TITLE
GlobusApp login, logout, and login_required

### LINKED ISSUE #1036: Add `GlobusApp` method to check login status
`GlobusApp` objects currently enable users to manually trigger a login flow via `app.run_login_flow()`, but not to check if a user already has valid tokens (i.e., logged in). I can accomplish this by interacting with the `app._validating_token_storage` object, but that is not a public interface, nor intuitive.

I propose creating an `app.is_logged_in()` method, or similar, which supports a `resource_server` argument. If the user does not specify a resource server, then the method could validate tokens for all associated resource servers.

In addition, we might consider creating an `app.ensure_logged_in()` method, or similar, which would simply run `app.run_login_flow()` if `app.is_logged_in()` returns `False`.

### PR DESCRIPTION
**Pull request by derek-globus**
Closes #1036 

## Added

-   Added ``login(...)``, ``logout(...)``, and ``login_required(...)`` to the
    experimental ``GlobusApp`` construct.

    -   ``login(...)`` initiates a login flow if:

        -   the current entity requires a login to satisfy local scope requirements or
        -   ``auth_params``/``force=True`` is passed to the method.

    -   ``logout(...)`` remove and revokes the current entity's app-associated tokens.

    -   ``login_required(...)`` returns a boolean indicating whether the app believes
        a login is required to satisfy local scope requirements.

Removed

-   Made ``run_login_flow`` private in the experimental ``GlobusApp`` construct.
    Usage sites should be replaced with either ``app.login()`` or
    ``app.login(force=True)``.
    -   **Old Usage**
 
            ```python
            app = UserApp("my-app", client_id="<my-client-id>")
            app.run_login_flow()
            ```
    -   **New Usage**
  
            ```python
            app = UserApp("my-app", client_id="<my-client-id>")
            app.login(force=True)
            ```

----
📚 Documentation preview 📚: https://globus-sdk-python--1041.org.readthedocs.build/en/1041/

### CODE REVIEW

File: src/globus_sdk/experimental/globus_app/_validating_token_storage.py [hunk 1]
```diff
@@ -72,7 +72,7 @@ def __init__(
         :param consent_client: An AuthClient to be used for consent polling. If omitted,
             dependent scope requirements are ignored during validation.
         """
-        self._token_storage = token_storage
+        self.token_storage = token_storage
```
Comment (derek-globus) [inline src/globus_sdk/experimental/globus_app/_validating_token_storage.py:75]: Potentially worth calling this out in the changelog as well. Wasn't sure whether that rose to that level

File: docs/experimental/examples/oauth2/globus_app.rst [hunk 2]
```diff
@@ -138,15 +138,20 @@ invocation of your requested method will proceed as expected.
...
+will be cached for future use.
 
-This method accepts a single optional parameter, ``auth_params``, where a caller
-may specify additional session-based auth parameters such as requiring the use of an
-MFA token or rendering with a specific message:
+This method accepts two optional keyword args:
+
+-   ``auth_params`` a collection of additional auth parameters to customize the login.
+    This allows for specifications such as, requiring that a user be logged in with an
+    MFA token or rendering the authorization webpage with a specific message.
```
Comment (derek-globus) [inline docs/experimental/examples/oauth2/globus_app.rst:152]: I'd love to link out to another doc here expanding on what can be passed but the two candidates I looked at were:
* [SDK Docs on GAREs](https://globus-sdk-python.readthedocs.io/en/stable/experimental/auth_requirements_errors.html) which don't enumerate the list of valid options.
* [Auth API Docs on Authorize Code Grant](https://docs.globus.org/api/auth/reference/#authorization_code_grant_preferred) which includes additional items beyond those that go strictly into a GAREData object.

File: docs/experimental/examples/oauth2/globus_app.rst [hunk 3]
```diff
@@ -138,15 +138,20 @@ invocation of your requested method will proceed as expected.
 Manually Running Login Flows
 ----------------------------
 
-While your app will automatically initiate and oversee auth flows as detected, sometimes
-an author may want to explicitly control when an authorization occurs. To manually
-trigger a login flow, call ``GlobusApp.run_login_flow(...)``. This will initiate an auth
-flow requesting new tokens based on the app's currently defined scope requirements, and
-caching the resulting tokens for future use.
+While your app will automatically initiate and oversee login flows as detected,
```
Comment (ada-globus) [inline docs/experimental/examples/oauth2/globus_app.rst:141]:
```suggestion
While your app will automatically initiate and oversee login flows when needed,
```

File: docs/experimental/examples/oauth2/globus_app.rst [hunk 4]
```diff
@@ -138,15 +138,20 @@ invocation of your requested method will proceed as expected.
...
+determines that any requirements across any resource servers are unmet. Resulting tokens
+will be cached for future use.
 
-This method accepts a single optional parameter, ``auth_params``, where a caller
-may specify additional session-based auth parameters such as requiring the use of an
-MFA token or rendering with a specific message:
+This method accepts two optional keyword args:
+
+-   ``auth_params`` a collection of additional auth parameters to customize the login.
+    This allows for specifications such as, requiring that a user be logged in with an
```
Comment (ada-globus) [inline docs/experimental/examples/oauth2/globus_app.rst:151]:
```suggestion
    This allows for specifications such as requiring that a user be logged in with an
```

File: docs/experimental/examples/oauth2/globus_app.rst [hunk 5]
```diff
@@ -138,15 +138,20 @@ invocation of your requested method will proceed as expected.
...
+current scope requirements against available tokens, initiating a login flow if it
+determines that any requirements across any resource servers are unmet. Resulting tokens
+will be cached for future use.
 
-This method accepts a single optional parameter, ``auth_params``, where a caller
-may specify additional session-based auth parameters such as requiring the use of an
-MFA token or rendering with a specific message:
+This method accepts two optional keyword args:
+
+-   ``auth_params`` a collection of additional auth parameters to customize the login.
```
Comment (ada-globus) [inline docs/experimental/examples/oauth2/globus_app.rst:150]:
```suggestion
-   ``auth_params``, a collection of additional auth parameters to customize the login.
```

File: docs/experimental/examples/oauth2/globus_app.rst [hunk 6]
```diff
@@ -138,15 +138,20 @@ invocation of your requested method will proceed as expected.
...
+    This allows for specifications such as, requiring that a user be logged in with an
+    MFA token or rendering the authorization webpage with a specific message.
+-   ``force``, a boolean flag instruct the app to run a login flow even if it believes
+    that no such login flow is required.
```
Comment (ada-globus) [inline docs/experimental/examples/oauth2/globus_app.rst:154]:
```suggestion
-   ``force``, a boolean flag instructing the app to perform a login flow
    regardless of whether it is required.
```

File: src/globus_sdk/experimental/globus_app/app.py [hunk 7]
```diff
@@ -224,8 +224,67 @@ def _initialize_authorizer_factory(self) -> None:
...
+        auth_params: GlobusAuthorizationParameters | None = None,
+        force: bool = False,
+    ) -> None:
+        """
+        Log an auth entity into the app, if needed, storing the resulting tokens.
+
+        A login flow will be is run if:
+            * `auth_params` are provided
+            * `force` is set to True
+            * `self.login_required()` evaluates to True
```
Comment (ada-globus) [inline src/globus_sdk/experimental/globus_app/app.py:239]:
```suggestion
            * ``auth_params`` are provided,
            * ``force`` is set to True, or
            * ``self.login_required()`` evaluates to True.
```

File: src/globus_sdk/experimental/globus_app/app.py [hunk 8]
```diff
@@ -224,8 +224,67 @@ def _initialize_authorizer_factory(self) -> None:
...
+    def login(
+        self,
+        *,
+        auth_params: GlobusAuthorizationParameters | None = None,
+        force: bool = False,
+    ) -> None:
+        """
+        Log an auth entity into the app, if needed, storing the resulting tokens.
+
+        A login flow will be is run if:
```
Comment (ada-globus) [inline src/globus_sdk/experimental/globus_app/app.py:236]:
```suggestion
        A login flow will be performed if:
```

File: src/globus_sdk/experimental/globus_app/app.py [hunk 9]
```diff
@@ -224,8 +224,67 @@ def _initialize_authorizer_factory(self) -> None:
...
+        :param auth_params: An optional set of authorization parameters to establish
+            requirements and controls for the login flow.
+        :param force: If True, run a login flow even if one is not necessary.
+        """
+        if auth_params or force or self.login_required():
+            self._run_login_flow(auth_params)
+
+    def login_required(self) -> bool:
+        """
+        Determine if a login flow is required to interact with resource servers under
```
Comment (ada-globus) [inline src/globus_sdk/experimental/globus_app/app.py:250]:
```suggestion
        Determine if a login flow will be required to interact with resource servers under
```

File: src/globus_sdk/experimental/globus_app/app.py [hunk 10]
```diff
@@ -224,8 +224,67 @@ def _initialize_authorizer_factory(self) -> None:
...
+        the current scope requirements.
+
+        This will return false if:
+            * access tokens have never been issued
+            * access tokens have been issued but are underscoped
+            * access tokens have expired and wouldn't be resolved with refresh tokens
```
Comment (ada-globus) [inline src/globus_sdk/experimental/globus_app/app.py:256]:
```suggestion
            * access tokens have never been issued,
            * access tokens have been issued but have insufficient consents, or
            * access tokens have expired and wouldn't be resolved with refresh tokens.
```

File: src/globus_sdk/experimental/globus_app/app.py [hunk 11]
```diff
@@ -224,8 +224,67 @@ def _initialize_authorizer_factory(self) -> None:
...
+        for resource_server in self._scope_requirements.keys():
+            try:
+                self.get_authorizer(resource_server, skip_error_handling=True)
+            except TokenValidationError:
+                return True
+        return False
+
+    def logout(self) -> None:
+        """
+        Log an auth entity out of the app.
```
Comment (ada-globus) [inline src/globus_sdk/experimental/globus_app/app.py:269]:
```suggestion
        Logout an auth entity from the app.
```

File: src/globus_sdk/experimental/globus_app/app.py [hunk 12]
```diff
@@ -224,8 +224,67 @@ def _initialize_authorizer_factory(self) -> None:
...
+
+        This will remove and revoke all tokens stored for the current app user.
+        """
+        # Revoke all tokens, removing them from the underlying token storage
+        for resource_server in self._scope_requirements.keys():
+            token_storage = self._validating_token_storage.token_storage
+            token_data = token_storage.get_token_data(resource_server)
```
Comment (ada-globus) [inline src/globus_sdk/experimental/globus_app/app.py:276]:
```suggestion
        # Revoke all tokens, removing them from the underlying token storage
        token_storage = self._validating_token_storage.token_storage
        for resource_server in self._scope_requirements.keys():
            token_data = token_storage.get_token_data(resource_server)
```

File: src/globus_sdk/experimental/globus_app/app.py [hunk 13]
```diff
@@ -261,18 +320,25 @@ def _auth_params_with_required_scopes(
...
+        skip_error_handling: bool = False,
+    ) -> GlobusAuthorizer:
         """
         Get a ``GlobusAuthorizer`` from the app's authorizer factory for a specified
         resource server. The type of authorizer is dependent on the app.
 
-        :param resource_server: the resource server the Authorizer will provide
-            authorization headers for
+        :param resource_server: the resource server over which the requested Authorizer
+            should  provide authorization headers for.
```
Comment (ada-globus) [inline src/globus_sdk/experimental/globus_app/app.py:334]:
```suggestion
        :param resource_server: the resource server for which the requested authorizer
            should provide authorization headers.
```

File: src/globus_sdk/experimental/globus_app/app.py [hunk 14]
```diff
@@ -224,8 +224,67 @@ def _initialize_authorizer_factory(self) -> None:
...
+                if token_data.refresh_token:
+                    self._login_client.oauth2_revoke_token(token_data.refresh_token)
+                token_storage.remove_token_data(resource_server)
+
+        # Invalidate any cached authorizers
+        self._authorizer_factory.clear_cache()
+
     @abc.abstractmethod
-    def run_login_flow(
+    def _run_login_flow(
```
Comment (ada-globus) [inline src/globus_sdk/experimental/globus_app/app.py:287]: This docstring is incomplete. I can't comment directly on the affected line to provide a suggestion

File: src/globus_sdk/experimental/globus_app/app.py [hunk 15]
```diff
@@ -224,8 +224,67 @@ def _initialize_authorizer_factory(self) -> None:
...
+        Log an auth entity into the app, if needed, storing the resulting tokens.
+
+        A login flow will be is run if:
+            * `auth_params` are provided
+            * `force` is set to True
+            * `self.login_required()` evaluates to True
+
+        :param auth_params: An optional set of authorization parameters to establish
+            requirements and controls for the login flow.
+        :param force: If True, run a login flow even if one is not necessary.
```
Comment (ada-globus) [inline src/globus_sdk/experimental/globus_app/app.py:243]:
```suggestion
        :param force: If True, perform a login flow even if one is not necessary.
```

File: src/globus_sdk/experimental/globus_app/app.py [hunk 16]
```diff
@@ -261,18 +320,25 @@ def _auth_params_with_required_scopes(
...
         """
         Get a ``GlobusAuthorizer`` from the app's authorizer factory for a specified
         resource server. The type of authorizer is dependent on the app.
 
-        :param resource_server: the resource server the Authorizer will provide
-            authorization headers for
+        :param resource_server: the resource server over which the requested Authorizer
+            should  provide authorization headers for.
+        :param skip_error_handling: If True, skip the configured token validation error
+            handler when a TokenValidationError is raised. Default: False.
```
Comment (ada-globus) [inline src/globus_sdk/experimental/globus_app/app.py:336]:
```suggestion
            handler when a ``TokenValidationError`` is raised. Default: False.
```

**Review by ada-globus — CHANGES_REQUESTED** [review]

File: src/globus_sdk/experimental/globus_app/app.py (hunk 14 above)
Comment (derek-globus) [reply src/globus_sdk/experimental/globus_app/app.py:287]: Copied the text over from `login`'s docstring.

File: src/globus_sdk/experimental/globus_app/app.py (hunk 7 above)
Comment (ada-globus) [reply src/globus_sdk/experimental/globus_app/app.py]: Looks like part of this change is missing? I can check others, but might be a little bit before I can swing back through.

Comment (derek-globus) [reply src/globus_sdk/experimental/globus_app/app.py]: Oh my mistake. I didn't see the list-form part of the suggestion here, just the "``" to replace "`" enveloping.

I can update to that style as well.

Comment (derek-globus) [reply src/globus_sdk/experimental/globus_app/app.py]: Updated in latest rev.

**Comment by derek-globus** [comment]
Rebased from main to resolve these changes against #1040

**Review by ada-globus — APPROVED** [review]
🙌

File: src/globus_sdk/experimental/globus_app/app.py [hunk 17]
```diff
@@ -227,16 +227,75 @@ def _initialize_authorizer_factory(self) -> None:
...
+        auth_params: GlobusAuthorizationParameters | None = None,
+        force: bool = False,
+    ) -> None:
+        """
+        Log an auth entity into the app, if needed, storing the resulting tokens.
+
+        A login flow will be performed if:
+            * ``auth_params`` are provided,
+            * ``force`` is set to True, or
+            * ``self.login_required()`` evaluates to True.
```
Comment (kurtmckee) [inline src/globus_sdk/experimental/globus_app/app.py:242]: In Markdown or RST (I don't remember which, or if both), indented bullet points don't render as expected.

I recommend keeping bullet points at the same indentation level as the surrounding text (but don't know with certainty if that applies to RST here).

Are all of these `or` conditions? I'd prefer to have `or` appear at the end of each line if so.

```suggestion
        A login flow will be performed if:

        * ``auth_params`` are provided, or
        * ``force`` is set to True, or
        * ``self.login_required()`` evaluates to True.
```

File: src/globus_sdk/experimental/globus_app/app.py [hunk 18]
```diff
@@ -227,16 +227,75 @@ def _initialize_authorizer_factory(self) -> None:
...
+
+        :param auth_params: An optional set of authorization parameters to establish
+            requirements and controls for the login flow.
+        :param force: If True, perform a login flow even if one is not necessary.
```
Comment (kurtmckee) [inline src/globus_sdk/experimental/globus_app/app.py:246]: I was under the impression that `.login_required()` might not perfectly know if login is required (I'm thinking of 30-minute timeouts), so it might be helpful to rephrase this sentence.

```suggestion
        :param force: If True, perform a login flow even if one doesn't appear to be necessary.
```

File: src/globus_sdk/experimental/globus_app/app.py [hunk 19]
```diff
@@ -227,16 +227,75 @@ def _initialize_authorizer_factory(self) -> None:
...
+
+    def login_required(self) -> bool:
+        """
+        Determine if a login flow will be required to interact with resource servers
+        under the current scope requirements.
+
+        This will return false if:
+            * access tokens have never been issued,
+            * access tokens have been issued but have insufficient scopes, or
+            * access tokens have expired and wouldn't be resolved with refresh tokens
```
Comment (kurtmckee) [inline src/globus_sdk/experimental/globus_app/app.py:259]:
```suggestion

        * access tokens have never been issued, or
        * access tokens have been issued but have insufficient scopes, or
        * access tokens have expired and wouldn't be resolved with refresh tokens
```

File: src/globus_sdk/experimental/globus_app/app.py [hunk 20]
```diff
@@ -264,18 +323,25 @@ def _auth_params_with_required_scopes(
...
+        *,
+        skip_error_handling: bool = False,
+    ) -> GlobusAuthorizer:
         """
         Get a ``GlobusAuthorizer`` from the app's authorizer factory for a specified
         resource server. The type of authorizer is dependent on the app.
 
-        :param resource_server: the resource server the Authorizer will provide
-            authorization headers for
+        :param resource_server: the resource server for which the requested Authorizer
```
Comment (kurtmckee) [inline src/globus_sdk/experimental/globus_app/app.py:336]: The parameter lines are sentence-cased throughout this PR.

```suggestion
        :param resource_server: The resource server for which the requested Authorizer
```

**Review by kurtmckee — CHANGES_REQUESTED** [review]

File: src/globus_sdk/experimental/globus_app/app.py (hunk 17 above)
Comment (derek-globus) [reply src/globus_sdk/experimental/globus_app/app.py:242]: On point 1, these render appropriately in the generated docs: ([generated docs](https://globus-sdk-python--1041.org.readthedocs.build/en/1041/experimental/globus_app.html#globus_sdk.experimental.globus_app.GlobusApp.login)).

Comment (derek-globus) [reply src/globus_sdk/experimental/globus_app/app.py:242]: For these two unioned conditional lists, I'm just going to rephrase it to not sound like a proper english sentence.


```
A login flow will be performed if any of the following hold true:
  * ``auth_params are provided.
  * ...
```

File: src/globus_sdk/experimental/globus_app/app.py (hunk 18 above)
Comment (derek-globus) [reply src/globus_sdk/experimental/globus_app/app.py:246]: Sure, I can update

File: src/globus_sdk/experimental/globus_app/app.py (hunk 17 above)
Comment (ada-globus) [reply src/globus_sdk/experimental/globus_app/app.py]: I think that's a good solution 👍

**Review by kurtmckee — APPROVED** [review]

**Review by sirosen — APPROVED** [review]
